File: DictoMiner/python/ANC.py


""" Parser for American National Corpus dataset """

import re
import sys

from Dictionary import Dictionary
from Word import Word


class ANC:
    fname = r'../data/ANC-token-count.txt';
    
    @staticmethod
    def get_dictionary( ):
        """ Build a Dictionary based on the American National Coprus data. """
        
        dicto = Dictionary()
        
        print 'Parsing ANC data...'
        i = 0;
        nLines = 240000
        
        # open file for reading
        with open(ANC.fname, 'r') as fid:    
            for line in fid:
                tokens = ANC.parse_line(line)
                if tokens is None:
                    continue
                
                # save data to list
                word = Word(tokens['word'], tokens['count'], tokens['prob'], i);
                dicto.add_word(word)
                
                # increment counter and show progress
                i = i + 1;
                progress = float(i) / float(nLines)
                if (progress % 0.05) < 1e-4:
                    sys.stdout.write("\r%2.2f%%" %(progress*100))
                    sys.stdout.flush()
                    
        print '\nDone.'
        
        return dicto

    @staticmethod
    def parse_line(line):
        """ Parse a single line of the ANC dataset. """
        
        tokens = line.split()
        # skip empty and malformed lines
        if len(tokens) == 3:
            data = {}
            data['word'] = tokens[0].upper()
            data['count'] = int(tokens[1])
            data['prob'] = float(tokens[2])
        else:
            data = None
            
        return data
    
if __name__ == "__main__":    
    dicto = ANC.get_dictionary()

File: DictoMiner/matlab/ANC.m
classdef ANC
    
    methods (Static)
        
        function dbWrds = tokens( )
            %ANC Summary of this function goes here
            %   Detailed explanation goes here
            
            
            % open file for reading
            fname = '\\scires1\audio\Boyett\_twenty\data\ANC\ANC-token-count.txt';
            fid = fopen(fname,'r');
            
            nLines = 240e3;
            
            % get all words in the list
            i = 1; dbWrds(nLines,1).index = -1;
            while ~feof(fid)
                % get and parse current line
                cline = fgetl(fid);
                if isempty(cline), continue; end
                ct = textscan(cline,'%s %d %f');
                
                % save data to db
                word = ct{1}{1};
                dbWrds(i,1).index = i;
                dbWrds(i).string = upper(word);
                dbWrds(i).length = length(word);
                dbWrds(i).count = ct{2};
                dbWrds(i).freq = ct{3};
                
                % increment counter and show progress
                i = i+1;
                progress.bar(i/nLines);
            end
            % close file
            fclose(fid);
            progress.bar(1);
            
            
            % reduce to proper size (last line is TOTAL data)
            dbWrds = dbWrds(1:i-2);
        end
        
    end
    
end
